src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Hover/RemoteHoverService.cs

```diff
@@ -0,0 +1,219 @@
+{
+    internal sealed class Factory : FactoryBase<IRemoteHoverService>
+    {
+        protected override IRemoteHoverService CreateService(in ServiceArgs args)
+            => new RemoteHoverService(in args);
+    }
+
+    private readonly IClientCapabilitiesService _clientCapabilitiesService = args.ExportProvider.GetExportedValue<IClientCapabilitiesService>();
+
+    protected override IDocumentPositionInfoStrategy DocumentPositionInfoStrategy => PreferAttributeNameDocumentPositionInfoStrategy.Instance;
+
+    public ValueTask<RemoteResponse<Hover?>> GetHoverAsync(
+        JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo,
+        JsonSerializableDocumentId documentId,
+        Position position,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetHoverAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<RemoteResponse<Hover?>> GetHoverAsync(
+        RemoteDocumentContext context,
+        Position position,
+        CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        if (!codeDocument.Source.Text.TryGetAbsoluteIndex(position, out var hostDocumentIndex))
+        {
+            return NoFurtherHandling;
+        }
+
+        var clientCapabilities = _clientCapabilitiesService.ClientCapabilities;
+
+        var positionInfo = GetPositionInfo(codeDocument, hostDocumentIndex, preferCSharpOverHtml: true);
+
+        if (positionInfo.LanguageKind is RazorLanguageKind.Html or RazorLanguageKind.Razor)
+        {
+            // Sometimes what looks like a html attribute can actually map to C#, in which case its better to let Roslyn try to handle this.
+            if (!DocumentMappingService.TryMapToGeneratedDocumentPosition(codeDocument.GetCSharpDocument(), positionInfo.HostDocumentIndex, out _, out _))
+            {
+                // Acquire from client capabilities
+                var options = HoverDisplayOptions.From(clientCapabilities);
+
+                var razorHover = await HoverFactory
+                    .GetHoverAsync(codeDocument, hostDocumentIndex, options, context.GetSolutionQueryOperations(), cancellationToken)
+                    .ConfigureAwait(false);
+
+                if (razorHover is null)
+                {
+                    return CallHtml;
+                }
+
+                // Ensure that we convert our Hover to a Roslyn Hover.
+                var resultHover = ConvertHover(razorHover);
+
+                return Results(resultHover);
+            }
+        }
+
+        var csharpDocument = codeDocument.GetCSharpDocument();
+        if (!DocumentMappingService.TryMapToGeneratedDocumentPosition(csharpDocument, positionInfo.HostDocumentIndex, out var mappedPosition, out _))
```

Comment:
This is the same as line 68, so we'll always map at least once, and sometimes twice. Perhaps a `bool canMapToCSharp = DocumentMappingService.Try...`, and use on line 65 and here?

Reply:
I made a larger change to simplify the logic. I realized that we didn't need to map to C# at all, since `GetPositionInfo(...)` would have done that for us.

Reply:
Much easier to read now, though it does appear there is no test coverage for component attributes to actually verify that it's working. Would be good to add one before merging.